### PR TITLE
added task_progress method

### PR DESCRIPTION
**Pull request by pwills11**
RDISCROWD- 3842

**Describe your changes**
added task progress method to filter tasks by task fields and task info fields

**Testing performed**
tested default functionality as well as task filter and task info filter

**Additional context**
Add any other context about your contribution here.

### CODE REVIEW

File: test/test_api/test_project_api.py [hunk 1]
```diff
@@ -1154,6 +1154,38 @@ def test_user_progress_guidelines_updated(self):
...
+        headers = [('Authorization', user.api_key)]
+        category = CategoryFactory.create()
+        res = self.app.get('/api/project?all=1&category_id=%s' % category.id, headers=headers)
+
+        taskruns = []
+        for task in tasks:
+            taskruns.extend(AnonymousTaskRunFactory.create_batch(2, task=task))
+
+        # check basic query without constraints to filter tasks  
+        res = self.app.get('/api/project/1/taskprogress', follow_redirects=True)
```
Comment (ywrsusan) [inline test/test_api/test_project_api.py:1173]:
```
res = self.app.get
(url = '/api/project/1/taskprogress?api_key=%s' % user.api_key)
```

**Review by XiChenn — APPROVED** [review]
lgtm

File: pybossa/api/__init__.py [hunk 2]
```diff
@@ -355,6 +359,63 @@ def user_progress(project_id=None, short_name=None):
...
+    sql_text = "SELECT COUNT(*) FROM task WHERE project_id=" + str(project.id)
+
+    task_info_fields = get_searchable_columns(project.id)
+    task_fields = [
+        "id",
+        "state",
+        "n_answers",
+        "created",
+        "calibration",
+    ]
```
Comment (dchhabda) [inline pybossa/api/__init__.py:396]: this can be outside of function

File: pybossa/api/__init__.py [hunk 3]
```diff
@@ -80,6 +80,10 @@
 from pybossa.auth.task import TaskAuth
 from pybossa.service_validators import ServiceValidators
 import requests
+from sqlalchemy.sql import text
+from pybossa.model.user import User
+from pybossa.core import db
+from pybossa.cache.task_browse_helpers import is_valid_searchable_column, get_searchable_columns
```
Comment (dchhabda) [inline pybossa/api/__init__.py:86]: is_valid_searchable_column is unused and can be removed

File: test/test_api/test_project_api.py [hunk 4]
```diff
@@ -1154,6 +1154,31 @@ def test_user_progress_guidelines_updated(self):
...
+        taskruns = []
+        for task in tasks:
+            taskruns.extend(AnonymousTaskRunFactory.create_batch(2, task=task))
+
+        # check basic query without constraints to filter tasks  
+        res = self.app.get('/api/project?all=1&category_id=%s' % category.id, headers=headers, follow_redirects=True)
+        assert res.status_code == 200
+
+        # check 404 response when the project doesn't exist   
+        res = self.app.get('/api/project//taskprogress', follow_redirects=True, headers=headers)
```
Comment (dchhabda) [inline test/test_api/test_project_api.py:1176]: cosmetic; extra / can be removed

File: pybossa/api/__init__.py [hunk 5]
```diff
@@ -80,6 +80,10 @@
 from pybossa.auth.task import TaskAuth
 from pybossa.service_validators import ServiceValidators
 import requests
+from sqlalchemy.sql import text
+from pybossa.model.user import User
```
Comment (dchhabda) [inline pybossa/api/__init__.py:84]: unused?

File: test/test_api/test_project_api.py (hunk 1 above)
Comment (pwills11) [reply test/test_api/test_project_api.py]: Thanks, using the header with the request worked for authorization!

File: pybossa/api/__init__.py [hunk 6]
```diff
@@ -355,6 +355,70 @@ def user_progress(project_id=None, short_name=None):
...
+@ratelimit(limit=ratelimits.get('LIMIT'), per=ratelimits.get('PER'))
+def task_progress(project_id=None, short_name=None):
+    """API endpoint for task progress.
+
+    Returns a JSON object continaing the number of tasks which meet the user defined filter constraints
+    """
+    from sqlalchemy.sql import text
+    from pybossa.model.user import User
+    from pybossa.core import db
+    from pybossa.cache.task_browse_helpers import is_valid_searchable_column, get_searchable_columns
```
[the remark this answers is not in the document]
Comment (pwills11) [reply pybossa/api/__init__.py]: done!

File: test/test_api/test_project_api.py (hunk 4 above)
Comment (pwills11) [reply test/test_api/test_project_api.py:1176]: which lines specifically are cosmetic?

File: pybossa/api/__init__.py (hunk 5 above)
Comment (pwills11) [reply pybossa/api/__init__.py:84]: I use both of these in the method task_prgress

File: pybossa/api/__init__.py (hunk 2 above)
Comment (pwills11) [reply pybossa/api/__init__.py:396]: should this go on the top of the file below headers? I'm not familiar when it's best to separate the list of fields in this case from the method.

File: pybossa/api/__init__.py (hunk 3 above)
Comment (pwills11) [reply pybossa/api/__init__.py:86]: great point. Thanks! I used it initially, but decided it was unnecessary

File: pybossa/api/__init__.py (hunk 5 above)
Comment (pwills11) [reply pybossa/api/__init__.py:84]: That's right. I will remove it

**Pull request merged**